pkg/k8s/controllers/reconciler.go

```diff
@@ -0,0 +1,276 @@
+}
+
+// Reconciler is a controller for CRD resources.
+type Reconciler struct {
+	// client accesses k8s api
+	client client.Client
+	// kind is the CRD kind served by this reconciler
+	kind string
+	// version is the version of CRD served by this controller
+	version string
+	// log preconfigured with reconciler fields.
+	log logrus.FieldLogger
+	// Handler is the reconciler's implementaion.
+	handler Handler
+}
+
+// reconcileResult holds the outcome of the reconciler
+type ReconcileResult struct {
+	// Skipped indicates that reconciler has decided to skip this CR.
+	Skipped bool
+	// ReconcileErr is set when reconciliation fails, or when status changes fail.
+	ReconcileErr error
+	// PropagateErr is set when reconciler error needs to be reported as a failure to the controller and retry immedaitely
+	// (rather than in intervals).
+	PropagateErr bool
+	// ControllerRes is the result to be returned back to the controller's infra.
+	ControllerRes ctrl.Result
+	// failCount is for internal use, holding number of times CR with the same hash has failed so far
+	failCount int
+}
+
+// NewReconciler creates a new reconciler instance.
+func NewReconciler(cl client.Client, kind, version string, log logrus.FieldLogger, handler Handler) *Reconciler {
+	return &Reconciler{
+		client:  cl,
+		kind:    kind,
+		version: version,
+		log:     log,
+		handler: handler,
+	}
+}
+
+// Kind returns the CRD's kind
+func (r *Reconciler) Kind() string {
+	return r.kind
+}
+
+// Version returns the CRD's version served by this reconciler
+func (r *Reconciler) Version() string {
+	return r.version
+}
+
+// Client returns the client to access k8s API.
+func (r *Reconciler) Client() client.Client {
+	return r.client
+}
+
+// endReconcile is invoked when Reconciler finishes.
+// This method is for logging and metrics and should NOT modify the ReconcileResult.
+func (r *Reconciler) endReconcile(
+	ctx context.Context, //nolint:unparam // Why: ctx might be ignored
+	log logrus.FieldLogger,
+	rr ReconcileResult,
+) {
+	// pass rr by value - EndReconcile should not tamper with the result
+	r.handler.EndReconcile(ctx, log, rr)
+
+	if rr.ReconcileErr != nil {
+		// make sure error messages are never lost
+		log.WithError(rr.ReconcileErr).Error("Reconciler failed to apply the CR")
+		if rr.PropagateErr {
+			log.Error("The error will be propageted to the controller")
+		}
+		return
+	}
+
+	if rr.Skipped {
+		log.Info("Reconciler skipped this event")
+		return
+	}
+
+	log.Info("Reconciler has applied the CR successfully")
+}
+
+// Reconcile is invoked when controller receives resource spec.
+func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := r.log.WithField("resourceName", req.NamespacedName).WithField("kind", r.Kind())
+	rr := r.doReconcile(ctx, log, req)
+
+	// defer as func to avoid 'capturing' rr - we want to provide latest values to the endReconcile
+	defer func() {
+		r.endReconcile(ctx, log, *rr)
```

Comment:
Strictly speaking, we don't actually need this callback setup at all.  Any service which uses this package can actually do their `post reconcile` activity by just embeddding:

```golang

type MyReconciler struct {
    // embed controller's Reconciler
    controllers.Reconciler
}
func (my *MyReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
    result, err := my.Reconciler.Reconciler(ctx, req)
    // do endReconcile work on top of result, err here.
}
```

I suggest just dropping EndReconcile out of the Handler interface altogether and this defer also.  If we really want to share the code of how to log the result, I suggest adding a helper to Reconciler to lot the result:

```golang
func (r *Reconciler) LogResult(ctx context.Context, req ctr.Result, res ctrl.Result, err error) {
}
```

Reply:
endReconcile carries skip logic - which is not available as a return value. the built-in version delegates to Handler.EndReconcile and also log full state, and it needs skip

Reply:
Maybe add skip to `Reconciler.Reconciler` ?  Another option is to return a custom error for skip.

Reply:
also, Handler's Reconcile is not always called - while EndReconcile is always invoked.

Reply:
Whether handler's Reconcile gets called or not doesn't really matter as I'm suggesting the bootstrapped service actually wraps `Reconciler.Reconcile` -- ie. wherever `controllers.NewReconciler()` is called, it would be replaced with `&MyReonciler{controllers.NewReconciler()}`.

Reply:
I could modify the Reconcile signature (make it sort of a helper utility) - but then controller must implement the interface. I wanted to keep bootstrapped code with 'code' in it to minimum, relying on Reconciler to be the primary receiver of the callback, while Handler gets all for free. EndReconcile is a way for the handler to handle success and error cases, and errors may come from the very beginning too (e.g. failed to even read CR from the k8s).

This way we can keep bootstrapped code to bare min, only asking for the actual core implementation.

Reply:
Isn't skip logic trivial to implement by returning `return ctrl.Result{}, nil` from Reconcile, I'm not sure. I concur with @vk-outreach, returning out of `Reconcile` is more consistent with the api and generated code from kubebuilder(which I fear we have strayed far from at this point).

Reply:
" return ctrl.Result{}, nil" is not skip - it is success, and we update LastReconcileSuccessHash on success. Skip means - do not update anything - just ignore the CR altogether because it belongs to a different bento or because it has been already applied.